spatial/r3/mat.go

```diff
@@ -0,0 +1,185 @@
+		j := iv % 3
+		M.data[i][j] = val[iv]
+	}
+	return &M
+}
+
+// Matrix interface implementation
+
+// Dims returns number of rows and columns of this matrix. This method will always return 3×3 for
+// r3.Mat type.
+func (m *Mat) Dims() (r, c int) { return 3, 3 }
+
+// At returns the element at the ith row and jth column. Expects indices
+// in the range [0,2]
+func (m *Mat) At(i, j int) float64 {
+	if uint(i) > 2 || uint(j) > 2 {
+		panic(badIdx)
+	}
+	return m.at(i, j)
+}
+
+func (m *Mat) at(i, j int) float64 {
+	return m.data[i][j]
+}
+
+// T returns the transpose of Mat.
+func (m *Mat) T() mat.Matrix { return mat.Transpose{Matrix: m} }
+
+// Set assigns v to the element at the ith row and jth column. Expects indices
+// in the range [0,2]
+func (m *Mat) Set(i, j int, v float64) {
+	if uint(i) > 2 || uint(j) > 2 {
+		panic(badIdx)
+	}
+	m.set(i, j, v)
+}
+
+// set sets element at position i,j without bounds checking
+func (m *Mat) set(i, j int, v float64) {
+	m.data[i][j] = v
+}
+
+// func (m *Mat) RawMatrix() blas64.General {
+// 	return blas64.General{Rows: 3, Cols: 3, Data: m[:], Stride: 3} // Can't easily perform RawMatrix without allocation or unsafe.
+// }
+
+// Eye returns the Identity 3×3 matrix
+func Eye() *Mat { return NewMat([]float64{1, 0, 0, 0, 1, 0, 0, 0, 1}) }
+
+// Scale multiplies the elements of a by f, placing the result in the receiver.
+//
+// See the Scaler interface for more information.
+func (m *Mat) Scale(f float64, a mat.Matrix) {
+	r, c := a.Dims()
+	if r != 3 || c != 3 {
+		panic(badDim)
+	}
+	for i := 0; i < 3; i++ {
+		for j := 0; j < 3; j++ {
+			m.set(i, j, f*a.At(i, j))
+		}
+	}
+}
+
+// Performs matrix multiplication on v:
+//  result = M * v
+func (m *Mat) MulVec(v Vec) (result Vec) {
+	result.X = v.X*m.at(0, 0) + v.Y*m.at(0, 1) + v.Z*m.at(0, 2)
+	result.Y = v.X*m.at(1, 0) + v.Y*m.at(1, 1) + v.Z*m.at(1, 2)
+	result.Z = v.X*m.at(2, 0) + v.Y*m.at(2, 1) + v.Z*m.at(2, 2)
+	return result
+}
+
+// Performs transposed matrix multiplication on v:
+//  result = Mᵀ * v
+func (m *Mat) MulVecTrans(v Vec) (result Vec) {
+	result.X = v.X*m.at(0, 0) + v.Y*m.at(1, 0) + v.Z*m.at(2, 0)
+	result.Y = v.X*m.at(0, 1) + v.Y*m.at(1, 1) + v.Z*m.at(2, 1)
+	result.Z = v.X*m.at(0, 2) + v.Y*m.at(1, 2) + v.Z*m.at(2, 2)
+	return result
+}
+
+// Skew returns the 3×3 skew symmetric matrix (right hand system) of v.
+//                  [0  -z  y]
+//  Skew({x,y,z}) = [z   0 -x]
+//                  [-y  x  0]
+func Skew(v Vec) (M *Mat) {
+	return NewMat([]float64{
+		0, -v.Z, v.Y,
+		v.Z, 0, -v.X,
+		-v.Y, v.X, 0,
+	})
+}
+
+// Mul takes the matrix product of a and b, placing the result in the receiver.
+// If the number of columns in a does not equal 3, Mul will panic.
+func (m *Mat) Mul(a, b mat.Matrix) {
+	// TODO(soypat): add optimized operations for Mat3*Mat3
+	// A, oka := a.(*Mat3)
+	// B, okb := b.(*Mat3)
+	// if oka && okb {
+	// 	m.mul(A, B)
+	// 	return
+	// }
+	d := mat.NewDense(3, 3, nil)
+	d.Mul(a, b)
+	m.CloneFrom(d)
+}
```

Comment:
https://go.dev/play/p/iUkCC5cF_1N
```
func (m *Mat) Mul(a, b mat.Matrix) {
	if r, c := a.Dims(); r != 3 || c != 3 {
		panic(badDim)
	}
	if r, c := b.Dims(); r != 3 || c != 3 {
		panic(badDim)
	}

	a00 = a.At(0, 0); b00 = b.At(0, 0)
	a01 = a.At(0, 1); b01 = b.At(0, 1)
	a02 = a.At(0, 2); b02 = b.At(0, 2)
	a10 = a.At(1, 0); b10 = b.At(1, 0)
	a11 = a.At(1, 1); b11 = b.At(1, 1)
	a12 = a.At(1, 2); b12 = b.At(1, 2)
	a20 = a.At(2, 0); b20 = b.At(2, 0)
	a21 = a.At(2, 1); b21 = b.At(2, 1)
	a22 = a.At(2, 2); b22 = b.At(2, 2)
	m.data[0][0] = a00*b00 + a01*b10 + a02*b20
	m.data[0][1] = a00*b01 + a01*b11 + a02*b21
	m.data[0][2] = a00*b02 + a01*b12 + a02*b22
	m.data[1][0] = a10*b00 + a11*b10 + a12*b20
	m.data[1][1] = a10*b01 + a11*b11 + a12*b21
	m.data[1][2] = a10*b02 + a11*b12 + a12*b22
	m.data[2][0] = a20*b00 + a21*b10 + a22*b20
	m.data[2][1] = a20*b01 + a21*b11 + a22*b21
	m.data[2][2] = a20*b02 + a21*b12 + a22*b22
}
```
With this, the type assertion to check whether `a` and `b` are `*Mat` probably doesn't pay (it might, but the direct arithmetic and 3× reduction is indexing is paying more already).

Reply:
Hmmmm, shouldn't we let the user multiply matrices which are not both necessarily 3x3?

Reply:
True. Make this the fast case when both a 3×3. Sorry.

Reply:
https://go.dev/play/p/L4c-TpDOcbJ

Reply:
That is a really neat way to generate matrix operations. It would not have occurred to me to actually do this. Thanks!